raspberry/SmartHome/app/src/test/java/smarthome/raspberry/data/HomeRepositoryImplTest.kt

```diff
@@ -0,0 +1,171 @@
+    : IotDevice(name, description, serveState, guid, controllers)
+
+abstract class B(name: String, description: String?,
+                 serveState: DeviceServeState, guid: Long,
+                 controllers: List<BaseController>)
+    : IotDevice(name, description, serveState, guid, controllers)
+
+
+class HomeRepositoryImplTest {
+
+    private lateinit var repo: HomeRepository
+    private val localStorage: LocalStorage = mock()
+    private val remoteStorage: RemoteStorage = mock()
+    private val deviceChannel_A: DeviceChannel = mock()
+    private val deviceChannel_B: DeviceChannel = mock()
+    private val device_A: A = mock()
+    private val device_B: B = mock()
+    private val controller: BaseController = mock()
+
+
+
+    @Test
+    fun hasController_proceedReadController_ReadStateFromChannel() {
+        runBlocking {
+            repo = HomeRepositoryImpl(localStorage, remoteStorage, mapOf(
+                    device_A.javaClass to deviceChannel_A
+            ))
+
+            whenever(localStorage.findDevice(controller)).then { device_A }
+            repo.proceedReadController(controller)
+            verify(deviceChannel_A).read(device_A, controller)
+        }
+    }
+
+    @Test
+    fun hasTwoChannels_proceedReadController_ReadStateFromProperChannel() {
+        runBlocking {
+            repo = HomeRepositoryImpl(localStorage, remoteStorage, mapOf(
+                    device_A.javaClass to deviceChannel_A,
+                    device_B.javaClass to deviceChannel_B
+            ))
+
+            whenever(localStorage.findDevice(any())).then { device_A }
+            repo.proceedReadController(controller)
+            verify(deviceChannel_A).read(device_A, controller)
+
+            whenever(localStorage.findDevice(any())).then { device_B }
+            repo.proceedReadController(controller)
+            verify(deviceChannel_B).read(device_B, controller)
+        }
+    }
+
+    @Test
+    fun hasTwoChannels_proceedWrite_WriteStateToProperChannel() {
+        runBlocking {
+            repo = HomeRepositoryImpl(localStorage, remoteStorage, mapOf(
+                    device_A.javaClass to deviceChannel_A,
+                    device_B.javaClass to deviceChannel_B
+            ))
+
+            val state = ControllerState()
+
+            whenever(localStorage.findDevice(any())).then { device_A }
+            repo.proceedWriteController(controller, state)
+            verify(deviceChannel_A).writeState(device_A, controller, state)
+        }
+    }
+
+    @Test
+    fun noDevices_findController_shouldThrow() {
+        runBlocking {
+            val repo: DeviceChannelOutput =
+                    HomeRepositoryImpl(localStorage, remoteStorage, mapOf(
+                            device_A.javaClass to deviceChannel_A
+                    ))
+
+            whenever(localStorage.getDevices()).thenReturn(mutableListOf())
+
+            shouldThrow<NoControllerException> {
+                repo.findController(123)
```

Comment:
DeviceChannelOutput does not have findController function

Reply:
why? 
`interface DeviceChannelOutput {
    suspend fun onNewDevice(device: IotDevice)
    suspend fun onNewState(controller: BaseController)
    suspend fun findController(guid: Long): BaseController
    suspend fun findDevice(controller: BaseController): IotDevice
}`